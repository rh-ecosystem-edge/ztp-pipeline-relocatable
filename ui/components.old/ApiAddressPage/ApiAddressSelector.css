.address-validation-failed {
  color: var(--pf-global--danger-color--100);
}
