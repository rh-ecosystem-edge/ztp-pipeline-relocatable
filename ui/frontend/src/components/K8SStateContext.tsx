import React from 'react';
import { isEqual } from 'lodash';

import {
  IpTripletSelectorValidationType,
  K8SStateContextData,
  K8SStateContextDataFields,
  CustomCertsValidationType,
} from './types';
import { ChangeDomainInputType, TlsCertificate } from '../copy-backend-common';
import {
  customCertsValidator,
  domainValidator,
  ipTripletAddressValidator,
  ipWithoutDots,
  passwordValidator,
  usernameValidator,
} from './utils';

const K8SStateContext = React.createContext<K8SStateContextData | null>(null);

export const K8SStateContextProvider: React.FC<{
  children: React.ReactNode;
}> = ({ children }) => {
  const [username, setUsername] = React.useState('');
  const [usernameValidation, setUsernameValidation] = React.useState<string>('');
  const handleSetUsername = React.useCallback(
    (newVal: string) => {
      setUsernameValidation(usernameValidator(newVal));
      setUsername(newVal);
    },
    [setUsername],
  );

  const [password, setPassword] = React.useState('');
  const [passwordValidation, setPasswordValidation] = React.useState(true);
  const handleSetPassword = React.useCallback(
    (newVal: string) => {
      setPasswordValidation(passwordValidator(newVal));
      setPassword(newVal);
    },
    [setPassword],
  );

  const [apiaddr, setApiaddr] = React.useState(ipWithoutDots('192.168.7.200')); // 192168  7200
  const [ingressIp, setIngressIp] = React.useState(ipWithoutDots('192.168.7.201')); // 192168  7201
  const [apiaddrValidation, setApiaddrValidation] = React.useState<IpTripletSelectorValidationType>(
    {
      valid: true,
      triplets: [],
    },
  );
  const handleSetApiaddr = React.useCallback(
    (newIp: string) => {
      setApiaddrValidation(ipTripletAddressValidator(newIp, ingressIp));
      setApiaddr(newIp);
    },
    [setApiaddr, ingressIp],
  );

  const [ingressIpValidation, setIngressIpValidation] =
    React.useState<IpTripletSelectorValidationType>({
      valid: true,
      triplets: [],
    });
  const handleSetIngressIp = React.useCallback(
    (newIp: string) => {
      setIngressIpValidation(ipTripletAddressValidator(newIp, apiaddr));
      setIngressIp(newIp);
    },
    [setIngressIp, apiaddr],
  );

  const [domain, setDomain] = React.useState<string>('');
  const [originalDomain, setOriginalDomain] = React.useState<string>();
  const [domainValidation, setDomainValidation] =
    React.useState<K8SStateContextData['domainValidation']>();
  const handleSetDomain = React.useCallback(
    (newDomain: string) => {
      setDomainValidation(domainValidator(newDomain));
      setDomain(newDomain);
      if (!originalDomain) {
        // Hint: This is expected to be called within initialDataLoad() only
        setOriginalDomain(newDomain);
      }
    },
    [originalDomain],
  );

  const [customCerts, setCustomCerts] = React.useState<ChangeDomainInputType['customCerts']>({});
  const [customCertsValidation, setCustomCertsValidation] =
    React.useState<CustomCertsValidationType>({});

  const setCustomCertificate = React.useCallback(
    (domain: string, certificate: TlsCertificate) => {
      const newCustomCerts = { ...customCerts };
      newCustomCerts[domain] = certificate;
      setCustomCerts(newCustomCerts);
      setCustomCertsValidation(customCertsValidator(customCertsValidation, domain, certificate));
    },
    [customCertsValidation, customCerts, setCustomCerts],
  );

  const isAllValid = React.useCallback(() => {
    const result =
      !usernameValidation &&
      passwordValidation &&
      apiaddrValidation.valid &&
      ingressIpValidation.valid &&
      !domainValidation &&
      !Object.keys(customCertsValidation).find(
        (d) =>
          customCertsValidation[d].certValidated === 'error' ||
          customCertsValidation[d].keyValidated === 'error',
      );
    return result;
  }, [
    apiaddrValidation.valid,
    customCertsValidation,
    domainValidation,
    ingressIpValidation.valid,
    passwordValidation,
    usernameValidation,
  ]);

  const _fv: K8SStateContextDataFields = {
    username,
    password,
    apiaddr,
    ingressIp,
    domain,
    originalDomain,
    customCerts,
  };

  const fieldValues = React.useRef<K8SStateContextDataFields>(_fv);
  fieldValues.current = _fv;

  const [snapshot, setSnapshot] = React.useState<K8SStateContextDataFields>(_fv);
  const setClean = React.useCallback(() => {
    setSnapshot(fieldValues.current);
  }, [fieldValues]);
  const isDirty = React.useCallback((): boolean => {
    return !isEqual(fieldValues.current, snapshot);
  }, [fieldValues, snapshot]);

  const value = {
    ...fieldValues.current,

    isDirty,
    setClean,
    isAllValid,

    usernameValidation,
    handleSetUsername,

    passwordValidation,
    handleSetPassword,

    apiaddrValidation,
    handleSetApiaddr,

    ingressIpValidation,
    handleSetIngressIp,

    domainValidation,
    handleSetDomain,

    customCertsValidation,
    setCustomCertificate,
  };

  return <K8SStateContext.Provider value={value}>{children}</K8SStateContext.Provider>;
};

export const useK8SStateContext = () => {
  const context = React.useContext(K8SStateContext);
  if (!context) {
    throw new Error('useK8SStateContext must be used within K8SStateContextProvider.');
  }
  return context;
};
