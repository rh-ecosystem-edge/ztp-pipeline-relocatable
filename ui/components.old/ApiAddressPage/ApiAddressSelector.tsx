import React from 'react';
import { Stack, StackItem, Title } from '@patternfly/react-core';

import { RequiredBadge } from '../Badge';
import { IpTripletsSelector } from '../IpTripletsSelector';
import { useK8SStateContext } from '../K8SStateContext';

import './ApiAddressSelector.css';

export const ApiAddressSelector: React.FC = () => {
  const { apiaddr, handleSetApiaddr, apiaddrValidation: validation } = useK8SStateContext();

  return (
    <Stack className="wizard-content" hasGutter>
      <StackItem>
        <Title headingLevel="h1">
          What is your API address? <RequiredBadge />
        </Title>
      </StackItem>
      <StackItem className="wizard-sublabel">
        Assign the IP address that will be used for API traffic.
      </StackItem>
      <StackItem>
        <IpTripletsSelector
          address={apiaddr}
          setAddress={handleSetApiaddr}
          validation={validation}
        />
      </StackItem>
      <StackItem isFilled>
        {validation.message && (
          <div data-testid="address-validation-failed" className="address-validation-failed">
            {validation.message}
          </div>
        )}
      </StackItem>
    </Stack>
  );
};
