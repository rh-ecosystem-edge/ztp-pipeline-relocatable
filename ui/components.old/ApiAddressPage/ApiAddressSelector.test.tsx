import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';

import { ApiAddressSelector } from './ApiAddressSelector';
import { K8SStateContextProvider, useK8SStateContext } from '../K8SStateContext';

const TestedComponent: React.FC<{
  ctxData?: { apiaddr: string /* apiaddrValidation: IpTripletSelectorValidationType*/ };
}> = ({ ctxData }) => {
  const { handleSetApiaddr } = useK8SStateContext();
  React.useEffect(() => {
    if (ctxData) {
      handleSetApiaddr(ctxData.apiaddr);
    }
  }, [ctxData, handleSetApiaddr]);

  return <ApiAddressSelector />;
};

const Component: React.FC<{
  ctxData?: { apiaddr: string };
}> = ({ ctxData }) => (
  <K8SStateContextProvider>
    <TestedComponent ctxData={ctxData} />
  </K8SStateContextProvider>
);

describe('ApiAddressSelector', () => {
  it('can render', () => {
    const { container } = render(<Component />);

    expect(screen.queryAllByTestId('address-validation-failed')).toHaveLength(0);
    expect(container).toMatchSnapshot();
  });

  it('can render non-default IP', () => {
    const { container } = render(<Component ctxData={{ apiaddr: '123 86111  2' }} />);

    expect(screen.getByTestId('ip-triplet-0')).toHaveValue('123');
    expect(screen.getByTestId('ip-triplet-1')).toHaveValue('86');
    expect(screen.getByTestId('ip-triplet-3')).toHaveValue('2');
    expect(screen.queryAllByTestId('address-validation-failed')).toHaveLength(0);

    expect(container).toMatchSnapshot();
  });

  it('can render validation error', () => {
    const { container } = render(<Component ctxData={{ apiaddr: '123586111  2' }} />);

    expect(screen.getByTestId('ip-triplet-0')).toHaveValue('123');
    expect(screen.getByTestId('ip-triplet-1')).toHaveValue('586');
    expect(screen.getByTestId('ip-triplet-3')).toHaveValue('2');

    expect(screen.queryAllByTestId('address-validation-failed')).toHaveLength(1);
    expect(screen.getByTestId('address-validation-failed')).toHaveClass(
      'address-validation-failed',
    );
    expect(screen.getByTestId('address-validation-failed')).toContainHTML(
      'Provided IP address is incorrect.',
    );

    expect(container).toMatchSnapshot();
  });

  it('can render change value', () => {
    const { container } = render(<Component ctxData={{ apiaddr: '123 86111  2' }} />);

    expect(screen.getByTestId('ip-triplet-0')).toHaveValue('123');
    expect(screen.getByTestId('ip-triplet-1')).toHaveValue('86');
    expect(screen.getByTestId('ip-triplet-3')).toHaveValue('2');

    const triplet0 = screen.getByTestId('ip-triplet-0');
    fireEvent.change(triplet0, { target: { value: '222' } });

    // chnage to correct value
    expect(screen.getByTestId('ip-triplet-0')).toHaveValue('222');
    expect(screen.getByTestId('ip-triplet-1')).toHaveValue('86');
    expect(screen.queryAllByTestId('address-validation-failed')).toHaveLength(0);

    // try to change to wrong value
    const triplet2 = screen.getByTestId('ip-triplet-2');
    fireEvent.change(triplet2, { target: { value: '555' } });
    // the value stays - component does not pass the wrong value
    expect(screen.getByTestId('ip-triplet-2')).toHaveValue('111');
    expect(screen.queryAllByTestId('address-validation-failed')).toHaveLength(0);

    expect(container).toMatchSnapshot();
  });
});
