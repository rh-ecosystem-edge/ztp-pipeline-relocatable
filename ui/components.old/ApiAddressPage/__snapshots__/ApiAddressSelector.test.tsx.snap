// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApiAddressSelector can render 1`] = `
<div>
  <div
    class="pf-l-stack pf-m-gutter wizard-content"
  >
    <div
      class="pf-l-stack__item"
    >
      <h1
        class="pf-c-title pf-m-2xl"
        data-ouia-component-id="OUIA-Generated-Title-1"
        data-ouia-component-type="PF4/Title"
        data-ouia-safe="true"
      >
        What is your API address? 
        <span
          class="required-badge"
        >
          *
        </span>
      </h1>
    </div>
    <div
      class="pf-l-stack__item wizard-sublabel"
    >
      Assign the IP address that will be used for API traffic.
    </div>
    <div
      class="pf-l-stack__item"
    >
      
      <input
        aria-invalid="false"
        aria-label="IP triplet, position 1"
        class="pf-c-form-control ip-triplet ip-triplet__wide"
        data-ouia-component-id="OUIA-Generated-TextInputBase-1"
        data-ouia-component-type="PF4/TextInput"
        data-ouia-safe="true"
        data-testid="ip-triplet-0"
        type="text"
        value="192"
      />
      .
      <input
        aria-invalid="false"
        aria-label="IP triplet, position 2"
        class="pf-c-form-control ip-triplet ip-triplet__wide"
        data-ouia-component-id="OUIA-Generated-TextInputBase-2"
        data-ouia-component-type="PF4/TextInput"
        data-ouia-safe="true"
        data-testid="ip-triplet-1"
        type="text"
        value="168"
      />
      .
      <input
        aria-invalid="false"
        aria-label="IP triplet, position 3"
        class="pf-c-form-control ip-triplet ip-triplet__wide"
        data-ouia-component-id="OUIA-Generated-TextInputBase-3"
        data-ouia-component-type="PF4/TextInput"
        data-ouia-safe="true"
        data-testid="ip-triplet-2"
        type="text"
        value="7"
      />
      .
      <input
        aria-invalid="false"
        aria-label="IP triplet, position 4"
        class="pf-c-form-control ip-triplet ip-triplet__wide"
        data-ouia-component-id="OUIA-Generated-TextInputBase-4"
        data-ouia-component-type="PF4/TextInput"
        data-ouia-safe="true"
        data-testid="ip-triplet-3"
        type="text"
        value="200"
      />
    </div>
    <div
      class="pf-l-stack__item pf-m-fill"
    />
  </div>
</div>
`;

exports[`ApiAddressSelector can render change value 1`] = `
<div>
  <div
    class="pf-l-stack pf-m-gutter wizard-content"
  >
    <div
      class="pf-l-stack__item"
    >
      <h1
        class="pf-c-title pf-m-2xl"
        data-ouia-component-id="OUIA-Generated-Title-4"
        data-ouia-component-type="PF4/Title"
        data-ouia-safe="true"
      >
        What is your API address? 
        <span
          class="required-badge"
        >
          *
        </span>
      </h1>
    </div>
    <div
      class="pf-l-stack__item wizard-sublabel"
    >
      Assign the IP address that will be used for API traffic.
    </div>
    <div
      class="pf-l-stack__item"
    >
      
      <input
        aria-invalid="false"
        aria-label="IP triplet, position 1"
        class="pf-c-form-control ip-triplet ip-triplet-success ip-triplet__wide"
        data-ouia-component-id="OUIA-Generated-TextInputBase-13"
        data-ouia-component-type="PF4/TextInput"
        data-ouia-safe="true"
        data-testid="ip-triplet-0"
        type="text"
        value="222"
      />
      .
      <input
        aria-invalid="false"
        aria-label="IP triplet, position 2"
        class="pf-c-form-control ip-triplet ip-triplet-success ip-triplet__wide"
        data-ouia-component-id="OUIA-Generated-TextInputBase-14"
        data-ouia-component-type="PF4/TextInput"
        data-ouia-safe="true"
        data-testid="ip-triplet-1"
        type="text"
        value="86"
      />
      .
      <input
        aria-invalid="false"
        aria-label="IP triplet, position 3"
        class="pf-c-form-control ip-triplet ip-triplet-success ip-triplet__wide"
        data-ouia-component-id="OUIA-Generated-TextInputBase-15"
        data-ouia-component-type="PF4/TextInput"
        data-ouia-safe="true"
        data-testid="ip-triplet-2"
        type="text"
        value="111"
      />
      .
      <input
        aria-invalid="false"
        aria-label="IP triplet, position 4"
        class="pf-c-form-control ip-triplet ip-triplet-success ip-triplet__wide"
        data-ouia-component-id="OUIA-Generated-TextInputBase-16"
        data-ouia-component-type="PF4/TextInput"
        data-ouia-safe="true"
        data-testid="ip-triplet-3"
        type="text"
        value="2"
      />
    </div>
    <div
      class="pf-l-stack__item pf-m-fill"
    />
  </div>
</div>
`;

exports[`ApiAddressSelector can render non-default IP 1`] = `
<div>
  <div
    class="pf-l-stack pf-m-gutter wizard-content"
  >
    <div
      class="pf-l-stack__item"
    >
      <h1
        class="pf-c-title pf-m-2xl"
        data-ouia-component-id="OUIA-Generated-Title-2"
        data-ouia-component-type="PF4/Title"
        data-ouia-safe="true"
      >
        What is your API address? 
        <span
          class="required-badge"
        >
          *
        </span>
      </h1>
    </div>
    <div
      class="pf-l-stack__item wizard-sublabel"
    >
      Assign the IP address that will be used for API traffic.
    </div>
    <div
      class="pf-l-stack__item"
    >
      
      <input
        aria-invalid="false"
        aria-label="IP triplet, position 1"
        class="pf-c-form-control ip-triplet ip-triplet-success ip-triplet__wide"
        data-ouia-component-id="OUIA-Generated-TextInputBase-5"
        data-ouia-component-type="PF4/TextInput"
        data-ouia-safe="true"
        data-testid="ip-triplet-0"
        type="text"
        value="123"
      />
      .
      <input
        aria-invalid="false"
        aria-label="IP triplet, position 2"
        class="pf-c-form-control ip-triplet ip-triplet-success ip-triplet__wide"
        data-ouia-component-id="OUIA-Generated-TextInputBase-6"
        data-ouia-component-type="PF4/TextInput"
        data-ouia-safe="true"
        data-testid="ip-triplet-1"
        type="text"
        value="86"
      />
      .
      <input
        aria-invalid="false"
        aria-label="IP triplet, position 3"
        class="pf-c-form-control ip-triplet ip-triplet-success ip-triplet__wide"
        data-ouia-component-id="OUIA-Generated-TextInputBase-7"
        data-ouia-component-type="PF4/TextInput"
        data-ouia-safe="true"
        data-testid="ip-triplet-2"
        type="text"
        value="111"
      />
      .
      <input
        aria-invalid="false"
        aria-label="IP triplet, position 4"
        class="pf-c-form-control ip-triplet ip-triplet-success ip-triplet__wide"
        data-ouia-component-id="OUIA-Generated-TextInputBase-8"
        data-ouia-component-type="PF4/TextInput"
        data-ouia-safe="true"
        data-testid="ip-triplet-3"
        type="text"
        value="2"
      />
    </div>
    <div
      class="pf-l-stack__item pf-m-fill"
    />
  </div>
</div>
`;

exports[`ApiAddressSelector can render validation error 1`] = `
<div>
  <div
    class="pf-l-stack pf-m-gutter wizard-content"
  >
    <div
      class="pf-l-stack__item"
    >
      <h1
        class="pf-c-title pf-m-2xl"
        data-ouia-component-id="OUIA-Generated-Title-3"
        data-ouia-component-type="PF4/Title"
        data-ouia-safe="true"
      >
        What is your API address? 
        <span
          class="required-badge"
        >
          *
        </span>
      </h1>
    </div>
    <div
      class="pf-l-stack__item wizard-sublabel"
    >
      Assign the IP address that will be used for API traffic.
    </div>
    <div
      class="pf-l-stack__item"
    >
      
      <input
        aria-invalid="false"
        aria-label="IP triplet, position 1"
        class="pf-c-form-control ip-triplet ip-triplet-success ip-triplet__wide"
        data-ouia-component-id="OUIA-Generated-TextInputBase-9"
        data-ouia-component-type="PF4/TextInput"
        data-ouia-safe="true"
        data-testid="ip-triplet-0"
        type="text"
        value="123"
      />
      .
      <input
        aria-invalid="false"
        aria-label="IP triplet, position 2"
        class="pf-c-form-control ip-triplet ip-triplet__wide"
        data-ouia-component-id="OUIA-Generated-TextInputBase-10"
        data-ouia-component-type="PF4/TextInput"
        data-ouia-safe="true"
        data-testid="ip-triplet-1"
        type="text"
        value="586"
      />
      .
      <input
        aria-invalid="false"
        aria-label="IP triplet, position 3"
        class="pf-c-form-control ip-triplet ip-triplet-success ip-triplet__wide"
        data-ouia-component-id="OUIA-Generated-TextInputBase-11"
        data-ouia-component-type="PF4/TextInput"
        data-ouia-safe="true"
        data-testid="ip-triplet-2"
        type="text"
        value="111"
      />
      .
      <input
        aria-invalid="false"
        aria-label="IP triplet, position 4"
        class="pf-c-form-control ip-triplet ip-triplet-success ip-triplet__wide"
        data-ouia-component-id="OUIA-Generated-TextInputBase-12"
        data-ouia-component-type="PF4/TextInput"
        data-ouia-safe="true"
        data-testid="ip-triplet-3"
        type="text"
        value="2"
      />
      <svg
        aria-hidden="true"
        class="validation-icon"
        fill="#c9190b"
        height="1em"
        role="img"
        style="vertical-align: -0.125em;"
        viewBox="0 0 512 512"
        width="1em"
      >
        <path
          d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zm-248 50c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
        />
      </svg>
    </div>
    <div
      class="pf-l-stack__item pf-m-fill"
    >
      <div
        class="address-validation-failed"
        data-testid="address-validation-failed"
      >
        Provided IP address is incorrect.
      </div>
    </div>
  </div>
</div>
`;
